Service status content type and related listing pages

stephen-cox opened the pull request:
This is the bulk of the work for the service status page as described here: https://trello.com/c/K7boYO9E/205-as-a-content-designer-create-and-manage-service-statuses

You can test this by pulling the https://github.com/localgovdrupal/localgov_project/tree/feature/13-service-status branch and doing the usual composer stuff.

All tests and coding standards checks for this module should pass.

There is a bug with the caching of landing page updates pages here: #28

stephen-cox commented:
Still to do is displaying the status on a landing page, but we'll handle this as part of the template stuff.

stephen-cox commented:
Hi @Adnan-cds, This does implement the Croydon way of doing things (https://new.croydon.gov.uk/service-status), but also the BHCC way (as described by @andybroomfield here: https://github.com/localgovdrupal/localgov_services/issues/13)

The BHCC way is there on the Croydon site, but isn't used; e.g. https://new.croydon.gov.uk/rubbish-and-recycling/update

Adnan-cds commented:
Okay, thanks Steve.

finnlewis commented:
Testing this locally now.

ekes commented:
Just chatting this one through:

I think we'd opt-ed for the use of 'Status' in place of 'Update'. It's mainly but not completely status at the moment. With some exception (Brighton's use) coming in for specific block/display where it is the latest 'Update'. I'd propose replacing all the Update use for Status now, and then making explicit exceptions?

finnlewis commented:
I agree @ekes that it would be good to get the consistency of naming convention in there now. 
This sounds like a fairly straightforward search / replace job and then re-test.

willguv commented:
I've probably missed the boat, but renaming everything "status" seems like a good idea :)

stephen-cox commented:
I have renamed the update stuff to status and fixed a minor issue I found with statuses being displayed on landing page status pages.

Ready to be reviewed again.

andybroomfield commented:
I have ran the tests sucessfully. 
BHCC has some more COVID work for me to do, I will pick up a review after I have done that.

ekes commented on tests/src/Functional/ServiceStatusTest.php:
```diff
@@ -0,0 +1,191 @@
+
+    // Rebuild caches.
+    // See: https://github.com/localgovdrupal/localgov_services/issues/28
+    drupal_flush_all_caches();
```
Is this not what `localgov_services_status_node_presave()` is supposed to do? Is it possibly not getting called in the test. Or is it not doing it's job?

ekes commented on modules/localgov_services_status/src/Routing/ServiceStatusRoutes.php:
```diff
@@ -0,0 +1,59 @@
+        ->execute();
+
+      if (!empty($status_nids)) {
+        $alias = \Drupal::service('path.alias_manager')->getAliasByPath('/node/' . $nid);
```
Not something that effects if this should be merged, but a note to ask everyone:

Are these sorts of path aliases of views more common? https://www.drupal.org/project/subpathauto is a nice install and forget solution to that.

ekes replied:
New issue about paths localgovdrupal/localgov#63

stephen-cox commented on modules/localgov_services_landing/localgov_services_landing.module:
```diff
@@ -25,6 +25,14 @@ function localgov_services_landing_preprocess_node(&$variables) {
       $variables['field_email_address'] = $node->get('field_email_address')->isEmpty() ? '' : $node->get('field_email_address')->first()->getValue()['value'];
       $variables['field_address_first_line'] = $node->get('field_address_first_line')->isEmpty() ? '' : $node->get('field_address_first_line')->first()->getValue()['value'];
       $variables['field_link_to_map'] = $node->get('field_link_to_map')->isEmpty() ? '' : $node->get('field_link_to_map')->first()->getValue()['uri'];
+
```
The commented out code will need to be rewritten as part of the displaying statuses on a landing page. I was planning on handling that with the theming of service pages.

ekes replied:
Probably remove it and pop it in the appropriate follow-up issue.

stephen-cox replied:
The whole hook function needs reviewing as part of the templating of the landing page. I was hoping to have done this in this sprint, but it will need to be done in the next sprint now. Until then I suggest we either remove the whole hook function or leave as is. 

I have created an issue for this sort of thing: https://github.com/localgovdrupal/localgov_services/issues/40

ekes commented:
https://github.com/localgovdrupal/localgov_services/issues/28#issuecomment-640208331

ekes commented on modules/localgov_services_status/config/install/core.entity_form_display.node.localgov_services_status.default.yml:
```diff
@@ -0,0 +1,117 @@
+    region: content
+    settings: {  }
+    third_party_settings: {  }
+  field_service:
```
This should now be fixed with https://github.com/localgovdrupal/localgov_services/pull/29/commits/2ff04f9bd9a6206e48aa80044c0dff2cd38c080e

andybroomfield commented:
Have been able to install and get service status running.
Verified that posting to /service-status checkbox works and that not checking them leaves them off the status page. As mentioned above, this does not add the status updates to the landing page as that functionality isn't here.

Only issue I've found is the order on the view at /service-status. Currently this is sticky and then a-z. Should be Revised -> Limited -> Normal and then a-z within them. Could be issue though with how you would order that in a view so may be worth changing the internal stored status labels to numbers to allow them to be ordered.

![Screenshot_2020-06-08 Service status localgovdrupal](https://user-images.githubusercontent.com/1467480/84006994-96814000-a967-11ea-97dc-f0d72f461aa0.png)

ekes commented:
>  Could be issue though with how you would order that in a view so may be worth changing the internal stored status labels to numbers to allow them to be ordered.

Looking at it I was worried it would be used for the css; which luckily it's not; but the machine names are used in some of the theme. Some maybe keeping string names but prefixing them with numbers: `0-severe-impact`, `1-has-issues`, `2-no-issues`?

andybroomfield commented:
Yes I noticed that. Wonder if we could refactor that bit.
Its used to mark up each label status in the template, and then css decides which to show / hide back on if that class name is present. When I migrated that to BHCC i replaced that with an if block in twig and could even be refactored further to just print the status label instead.

````twig
              {% if entity_service_status == 'severe-impact' %}
                  <small class="service-status-failure bhcc-alert bhcc-alert-borderless bhcc-alert-small bhcc-alert-failure">
                    {{ revised_service_label|t }}
                  </small>
                {% elseif entity_service_status == 'has-issues' %}
                  <small class="service-status-warning bhcc-alert bhcc-alert-borderless bhcc-alert-small bhcc-alert-warning">
                    {{ limited_service_label|t }}
                  </small>
                {% elseif entity_service_status == 'no-issues' %}
                  <small class="service-status-success bhcc-alert bhcc-alert-borderless bhcc-alert-small bhcc-alert-success">
                    {{ normal_service_label|t }}
                  </small>
                {% else %}
                  &nbsp;
                {% endif %}
````

ekes commented:
Reviewing everything I think deciding this field is the last thing before merging this.

Other issues: - formatting of the {landingpage}/status page, and adding a bloc to {landingpage} with the latest status and link to aforementioned page can be done as separate issues, as they're not structural to the status content type or anything else here.

stephen-cox commented:
> > Could be issue though with how you would order that in a view so may be worth changing the internal stored status labels to numbers to allow them to be ordered.
> 
> Looking at it I was worried it would be used for the css; which luckily it's not; but the machine names are used in some of the theme. Some maybe keeping string names but prefixing them with numbers: `0-severe-impact`, `1-has-issues`, `2-no-issues`?

The [Views List Sort](https://www.drupal.org/project/views_list_sort) module will sort views results by the order of the list items, but when I tried it the tests threw schema errors, so it might be more hassle than it's worth to use as this seems a much simpler fix.

stephen-cox commented:
The recent changes all seem good to me and work well. I have fixed a few minor coding standards errors.

There is a failing test, but this also happens on the master branch so will create a separate issue for that and look to fix it.

```
Drupal\Tests\localgov_services\Functional\ServicesBlock
 ✘ Service core blocks display
   │
   │ Behat\Mink\Exception\ExpectationException: The string "href="/node/1"" was not found anywhere in the HTML response of the current page.
   │ 
   │ /app/vendor/behat/mink/src/WebAssert.php:768
   │ /app/vendor/behat/mink/src/WebAssert.php:324
   │ /app/web/core/tests/Drupal/Tests/WebAssert.php:429
   │ /app/web/modules/contrib/localgov_services/tests/src/Functional/ServicesBlockTest.php:82
   │ 

Time: 1.23 minutes, Memory: 8.00 MB
```

stephen-cox commented:
Pull request for the failing tests is here: https://github.com/localgovdrupal/localgov_services/pull/42

ekes commented:
So with the follow-up #43 for the landing page. We're good.
